checker/checker.go

```diff
@@ -308,16 +307,20 @@ func (c *checker) resolveOverload(
 				checkedRef.OverloadId = append(checkedRef.OverloadId, overload.OverloadId)
 			}
 
+			// First matching overload, determines result type.
+			fnResultType := substitute(c.mappings,
+				overloadType.GetFunction().ResultType,
+				false)
 			if resultType == nil {
-				// First matching overload, determines result type.
-				resultType = substitute(c.mappings,
-					overloadType.GetFunction().ResultType,
-					false)
+				resultType = fnResultType
+			} else if !isDyn(resultType) &&
+				!isTypeParam(resultType) &&
+				!isTypeParam(fnResultType) &&
+				c.isAssignable(fnResultType, resultType) {
```

Comment:
Consider whether each of the checks in lines 316-319 are necessary.  Might the mostGeneral() function return the right result without them?  If they are necessary, please include a test case which would fail without having the check.

Reply:
Unfortunately, `mostGeneral` relies on the types already being checked for assignability. I did originally just check for type equality which covers a wide range of cases, but probably not the maximal set. Given the complexity of the condition though, I'm almost inclined to revert back to simple equality as the condition for not marking the return `dyn`. Thoughts?

I had originally omitted the `isTypeParam` check and this expression in the `checker_test.go` file is incorrectly tagged as an `int` when it should be a `dyn` since the assignability check adds a type substitution variable where there shouldn't be one: `x.single_value[23] + x.single_struct['y']`. Both `single_value` and `single_struct` are type `dyn`.

Reply:
The change looks simpler and more clearly correct, so yeah, go with that.